rfcs/51-animation-composition.md

```diff
@@ -0,0 +1,434 @@
+   from the associated graph. Fail if not found.
+ - The curves for the active clips is retrieved from the track's sparse array of
+   curves.
+ - Individual values are sampled from each active curve. Skipping curves with a
+   weight of 0.
+ - `Animatable::blend` is used to blend the sampled values together based on the
+   weights in the influences map.
+
+This approach has a number of performance related benefits:
+
+ - The state of the graph does not need to be evaluated for every sampling.
+ - Curves can be sampled in `ClipId` order based on the influence map. This
+   iteration should be cache friendly.
+ - No allocation is inherently required during sampling.
+ - This process is (mostly) branch free and can be accelerated easily with SIMD
+   compatible `Animatable::blend` implementations.
+
+Rough pseudo-code for sampling values with an exclusive system.
+
+```rust
+fn sample_animators(
+  animation_graphs: Query<(Entity, &AnimationGraph)>,
+  mut transforms: Query<&mut Transform>,
+  children: Query<&Children>,
+) {
+    animation_graphsgraphs.par_iter().for_each(|(root, graph)| {
+        let influences = graph.influences();
+        for (entity_path, bone) in graph.bones().iter() {
+            let Some(transform) = get_descendant(root, entity_path,
+                transforms, children) else { continue };
+            for track in bone.tracks() {
+                apply_track(*transform, track, influences);
+            }
+        }
+    });
+}
+
+fn apply_track(
+  transform: &mut Transform,
+  track: &Track,
+  influences: &GraphInfluences,
+) {
+  match track {
+      Track::Translation(ref curves) => {
+          let clips = influences.clips().iter();
+          let curves = track.curves().iter();
+          let blend_inputs = curves.zip(clips).map(|(curve, clip)| {
+              BlendInput {
+                  value: curve.sample(clip.time),
+                  weight: clip.weight,
+              }
+          });
+          if let Some(blend_result) = Vec3A::blend(blend_inputs) {
+              transform.translation = blend_result;t dqt ;w
+          }
+      },
+      Track::Scale(...) => {
+          ...
+      },
+      Track::Rotation(...) => {
+          ...
+      },
+  }
+}
+```
+
+## Drawbacks
+The animation sampling system is an exclusive system and blocks all other systems
+from running.
+
+## Rationale and alternatives
+
+### Relatonal `BoneBinding` as a Component
+Instead of using `BoneBindings` as a resource that is continually rebuilt every
+frame it's changed, `BoneBinding` could be a relational component, much like
+`Parent` or `Children`. This removes the need to scan the named hierarchy every
+frame, and allows trivial parallel iteration via `Query::par_for_each`. However:
+
+ - it does not avoid the need for userspace `unsafe`.
+ - maintaining the binding components is going to be a nightmare if there are
+   any hierarchy or name changes underneath an `AnimationGraph`.
```

Comment:
The current system has a path cache right? Why would that be impossible to establish in the graph-based system? How is maintaining the cache less of a nightmare? It should have the same limitations.

Reply:
The path cache is just a cache. It's not treated as a source of truth, just a opportunistic way to reduce  expensive searching. The binding mentioned here is more akin to relations and would have to be treated as a source of truth to get the full benefits, and that has the same mire of edge cases are still being worked out for the hierarchy.